src/null-safety/index.md

```diff
@@ -1,39 +1,257 @@
 ---
 title: Null safety
-description: Watch this space for news about null safety
-toc: false
+description: Information about Dart's upcoming null safety feature
 ---
 
-Watch this page for updates about null safety.
-In the meantime, check out these resources:
+Sound null safety is coming to the Dart language!
+When you opt into null safety,
+types in your code are non-nullable by default, meaning that
+values can’t be null _unless you say they can be._
+With null safety, your **runtime** null-dereference errors
+turn into **edit-time** analysis errors.
+The compiler can optimize away internal null checks,
+enabling apps to be faster, smaller, and more reliable.
 
-* [Null-safe playground: nullsafety.dartpad.dev][nullsafety.dartpad.dev]
+You can practice using null safety in the web app
+[DartPad with Null Safety.][nullsafety.dartpad.dev]
+
+{{ site.alert.important }}
+  **Don't start migrating large projects yet.**
+  Do test the feature and [give us feedback.][]
+{{ site.alert.end }}
+
+[give us feedback.]: https://github.com/dart-lang/sdk/issues/new?title=Null%20safety%20feedback:%20[issue%20summary]&labels=NNBD&body=Describe%20the%20issue%20or%20potential%20improvement%20in%20detail%20here
+
+If you've used Kotlin, TypeScript, or C#,
+then some of the null-safety syntax might look familiar.
+That's by design: the Dart language aims to be unsurprising.
+
+
+## Creating and using variables (?, !, late)
+
+Use `?`, `!`, and `late` to tweak how the Dart analyzer treats your code.
+
+Here’s an example of declaring a **non-nullable integer variable**
+(assuming you’ve opted into null safety):
+
+```dart
+int aRealInt; // Can't be used until it's initialized to an integer.
+```
+
+If the variable _can_ have the value `null`,
+**add `?`** to its type declaration:
+
+```dart
+int? aNullableInt = null;
+```
+
+When using a value that you know isn’t null but that has a nullable type,
+**add `!`** to tell the Dart analyzer that the value isn’t null:
```

Comment:
Again, it's not just the Dart analyzer.

Reply:
Also, should we recommend doing a null check instead? Some guidance on when to use the (potentially bad) bang operator would be good.

Reply:
For comparison, see how the Kotlin docs treats their `!` (which is `!!`) operator here: https://kotlinlang.org/docs/reference/null-safety.html

Reply:
I actually think the situation where you *know* that the value is non-null, but the type system can't see that (say, you know that when `state` is `running`, the `runningTask` is non-null, and you just checked `state`) is *exactly* the place to use `!`.

Reply:
Sure, but here it's very early in our docs, and we've not described the other ways yet (like actually testing for null and relying on flow analysis), so someone new to null safety might get the idea that `!` is needed for any access on a nullable variable...

Reply:
I've added a note and example about explicitly checking for null if you aren't sure about the value.